File: examples/chapter06/LIAF.m
%%
% Consider the ridiculously simple IVP $u'=u$, $u(0)=1$, whose solution is
% $e^t$. 
dudt = @(t,u) u;
u_exact = @exp;
a = 0;  b = 1;

%%
% Let's apply the LIAF method to this problem for varying fixed step
% sizes. We'll measure the error at the time $t=1$. 
n = [5;10;20;40;60];
err = 0*n;
for j = 1:length(n)
    h = (b-a)/n(j);
    t = a + h*(0:n(j))';
    u = [1; u_exact(h); zeros(n(j)-1,1)];
    f = [dudt(t(1),u(1)); zeros(n(j)-2,1)];
    for i = 2:n(j)
        f(i) = dudt(t(i),u(i));
        u(i+1) = -4*u(i) + 5*u(i-1) + h*(4*f(i) + 2*f(i-1));
    end
    err(j) = abs(u_exact(b) - u(end));
end

h = (b-a)./n;
table(n,h,err)
    
%%
% The error starts out promisingly, but things explode from there. A graph
% of the last numerical attempt yields a clue.
semilogy(t,abs(u),'.-')
xlabel('t'),  ylabel('|u|')  % ignore this line
title('LIAF solution'), axis tight   % ignore this line

%%
% It's clear that the solution is growing exponentially in time.

File: julia/scripts/chapter06/LIAF.jl

dudt = (u,t) -> u;
u_exact = exp;
a = 0.0;  b = 1.0;

n = [5,10,20,40,60]
err = zeros(size(n))
t = [];  u = [];
for (j,n) = enumerate(n)
    h = (b-a)/n
    t = [ a + i*h for i=0:n ]
    u = [1; u_exact(h); zeros(n-1)];
    f = [dudt(u[1],t[1]); zeros(n)];
    for i = 2:n
        f[i] = dudt(u[i],t[i])
        u[i+1] = -4*u[i] + 5*u[i-1] + h*(4*f[i]+2*f[i-1])
    end
    err[j] = abs(u_exact(b) - u[end])
end

using DataFrames
DataFrame(n=n,h=(b-a)./n,error=err)    

using Plots
plot(t,abs.(u),m=:o,label="",
    xlabel="t",yaxis=(:log10,"|u|"),title="LIAF solution")
